src/node_crypto.cc

```diff
@@ -2097,6 +2099,42 @@ void SSLWrap<Base>::GetPeerCertificate(
 }
 
 
+template <class Base>
+void SSLWrap<Base>::GetFinished(const FunctionCallbackInfo<Value>& args) {
+  Environment* env = Environment::GetCurrent(args);
+
+  Base* w;
+  ASSIGN_OR_RETURN_UNWRAP(&w, args.Holder());
+
+  char* buf = Malloc(EVP_MAX_MD_SIZE * 2);
+  size_t len = SSL_get_finished(w->ssl_, buf, EVP_MAX_MD_SIZE * 2);
+  if (!len) {
```

Comment:
I think it might be worth verifying that `len` is less than `EVP_MAX_MD_SIZE * 2`.

Reply:
@indutny This test is done in [the routines themselves][1].

[1]: https://github.com/nodejs/node/blob/master/deps/openssl/openssl/ssl/ssl_lib.c#L795-L821

Reply:
Yes, it is done there. However, what I'm asking to check is that the data is complete and not partial.

Reply:
@indutny Oh, I see now what you mean. In order to ensure that the message is not being cut off, I provide the `buf`fer of the same size that [the corresponding arrays][1] have in `<openssl/ssl3.h>`. Also, the comment there suggests that `EVP_MAX_MD_SIZE * 2` is an overstatement:
```
        /* actually only need to be 16+20 for SSLv3 and 12 for TLS */
        unsigned char finish_md[EVP_MAX_MD_SIZE * 2];
        int finish_md_len;
        unsigned char peer_finish_md[EVP_MAX_MD_SIZE * 2];
        int peer_finish_md_len;
```

[1]: https://github.com/nodejs/node/blob/master/deps/openssl/openssl/include/openssl/ssl3.h#L556-L560

Reply:
`assert` then?

Reply:
> `assert` then?

@indutny Could you please elaborate?

Reply:
I suggest to add an `assert` here anyway. We've little control over OpenSSL codebase, and it'd be great to have a check for partial data even if such can't be returned right now.

Reply:
> I suggest to add an assert here anyway. We've little control over OpenSSL codebase, and it'd be great to have a check for partial data even if such can't be returned right now.

@indutny I am sorry, but I still cannot understand what assertion exactly do you mean?

Reply:
@indutny Nevermind, I amended the commit to call `SSL_get_finished` and `SSL_get_peer_finished` twice: first just to get the actual `len`gth, then to copy the message to the `buf`fer.

Reply:
Unfortunately, this is undefined behavior:

```
If an argument to a function has an invalid value (such as a value outside the domain of the function, or a pointer outside the address space of the program, or a null pointer, or a pointer to non-modifiable storage when the corresponding parameter is not const-qualified) or a type (after promotion) not expected by a function with variable number of arguments, the behavior is undefined.
```

See: https://www.imperialviolet.org/2016/06/26/nonnull.html

Reply:
I'm fine with what it looked like before. Could you please revert it to that form and add:

```C
CHECK_LT(len, EVP_MAX_MD_SIZE * 2);
```

Reply:
@indutny Oops, I did not know about section 7.1.4, my bad. Still, I would prefer avoiding the `EVP_MAX_MD_SIZE * 2` constant because it is unclear how exactly the latter is related to the `Finished` messages. So, I further amended the commit to use a `dummy` array instead of `nullptr` not to cause undefined behavior in `memcpy()`.